core/src/main/java/org/apache/iceberg/PartitionsTable.java

```diff
@@ -96,27 +94,63 @@ private static StaticDataTask.Row convertPartition(Partition partition) {
   }
 
   private static Iterable<Partition> partitions(Table table, StaticTableScan scan) {
-    CloseableIterable<FileScanTask> tasks = planFiles(scan);
     Types.StructType normalizedPartitionType = Partitioning.partitionType(table);
     PartitionMap partitions = new PartitionMap();
 
     // cache a position map needed by each partition spec to normalize partitions to final schema
     Map<Integer, int[]> normalizedPositionsBySpec =
         Maps.newHashMapWithExpectedSize(table.specs().size());
 
-    for (FileScanTask task : tasks) {
-      PartitionData original = (PartitionData) task.file().partition();
-      int[] normalizedPositions =
-          normalizedPositionsBySpec.computeIfAbsent(
-              task.spec().specId(),
-              specId -> normalizedPositions(table, specId, normalizedPartitionType));
+    int[] normalizedPositions =
+        normalizedPositionsBySpec.computeIfAbsent(
+            table.spec().specId(),
+            specId -> normalizedPositions(table, specId, normalizedPartitionType));
+
+    CloseableIterable<DataFile> datafiles = planDataFiles(scan);
+
+    for (DataFile dataFile : datafiles) {
+      PartitionData original = (PartitionData) dataFile.partition();
       PartitionData normalized =
           normalizePartition(original, normalizedPartitionType, normalizedPositions);
-      partitions.get(normalized).update(task.file());
+      partitions.get(normalized).update(dataFile);
     }
+
     return partitions.all();
   }
 
+  @VisibleForTesting
+  static CloseableIterable<DataFile> planDataFiles(StaticTableScan scan) {
```

Comment:
I know we are just refactoring the current logic. So, `DataFile` is fine. 

But to avoid back-and-forth changes (when we support delete file stats), Maybe we need to just call it `PlanFiles` and return `CloseableIterable<ContentFile> `

Reply:
Yea I don't see an easy way in this pr , as there are going to be a few changes, let's do it in that pr right after this one then.

Reply:
Thank you @ajantha-bhat. I guess as Szehon mentioned, to read delete files I guess majority of code can be reused by it would be `ManifestReader.readDeleteManifest()` instead and I assume the file count and file size will be aggregated differently so I didn't move this to ContentFile and keep the scope as refactoring.

Reply:
ok. SGTM.